include/ensmallen_bits/nelder_mead/nelder_mead_impl.hpp

```diff
@@ -0,0 +1,241 @@
+  // at the vertices. order(0) is the index in the simplex of the vertex
+  // with the lowest function value, and order(dim) is the index in the
+  // simplex of the vertex with the highest function value.
+  arma::uvec order = arma::sort_index(fx);
+
+  terminate |= Callback::BeginOptimization(*this, function, iterate,
+      callbacks...);
+  for (size_t i = 1; i != maxIterations && !terminate; ++i)
+  {
+    const MatType M = arma::mean(simplex.cols(order.subvec(0, dim - 1)), 1);
+    const MatType xref = (1.0 + alpha) * M - alpha * simplex.col(order(dim));
+    const ElemType fref = function.Evaluate(xref);
+    Callback::Evaluate(*this, function, xref, fref, callbacks...);
+
+    if (fref < fx(order(0)))
+    {
+      // Compute expansion.
+      const MatType xexp = (1.0 + alpha * beta) * M - alpha * beta *
+          simplex.col(order(dim));
+      const ElemType fexp = function.Evaluate(xexp);
+      Callback::Evaluate(*this, function, xexp, fexp, callbacks...);
+
+      // Shift values by one to the right and swap the last with the first.
+      const size_t fLowest = order(dim);
+      for (size_t d = dim; d > 0; --d)
+        order(d) = order(d - 1);
+      order(0) = fLowest;
+
+      if (fexp < fref)
+      {
+        simplex.col(order(0)) = xexp;
+        fx(order(0)) = fexp;
+      }
+      else
+      {
+        simplex.col(order(0)) = xref;
+        fx(order(0)) = fref;
+      }
+    }
+    else if (fref < fx(order(dim - 1)))
+    {
+      // Accept reflection point.
+      const size_t s = ShiftOrder(order, fx, fref);
+      simplex.col(order(s)) = xref;
+      fx(order(s)) = fref;
+    }
+    else
+    {
+      if (fref < fx(order(dim)))
+      {
+        // Outside contraction.
+        const MatType xoc = (1.0 + alpha * gamma) * M - alpha * gamma *
+            simplex.col(order(dim));
+        const ElemType foc = function.Evaluate(xoc);
+        Callback::Evaluate(*this, function, xoc, foc, callbacks...);
+
+        if (foc <= fref)
+        {
+          const size_t s = ShiftOrder(order, fx, fref);
+          simplex.col(order(s)) = xoc;
+          fx(order(s)) = foc;
+        }
+        else
+        {
+          shrink = true;
+        }
+      }
+      else
+      {
+        // Inside contraction.
+        const MatType xic = (1.0 - gamma) * M + gamma *
+            simplex.col(order(dim));
+        const ElemType fic = function.Evaluate(xic);
+        Callback::Evaluate(*this, function, xic, fic, callbacks...);
+
+        if (fic < fx(order(dim)))
+        {
+          const size_t s = ShiftOrder(order, fx, fic);
+          simplex.col(order(s)) = xic;
+          fx(order(s)) = fic;
+        }
+        else
+        {
+          shrink = true;
+        }
+      }
+    }
+
+    if (shrink)
+    {
+      for (size_t d = 1; d < dim; ++d)
+      {
+        simplex.col(order(d)) = simplex.col(order(0)) + delta *
+            (simplex.col(order(d)) - simplex.col(order(0)));
```

Comment:
It could be possible to express this as two operations (first on columns with index less than `order(0)`, and second on columns with index greater than `order(0)`), using `repmat()`, but it's up to you if you want to rewrite it like that.  I believe it would be more efficient, but I don't know how much of a bottleneck it would be.  The `Evaluate()` would still need to be called over all new points in a loop though.